dnsclient/dnsclient.go

```diff
@@ -0,0 +1,195 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dnsclient is a client library for performing CT operations
+// over DNS.  The DNS mechanism is experimental and subject to change.
+package dnsclient
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base32"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/tls"
+)
+
+// DNSClient represents a DNS client for a given CT Log instance
+type DNSClient struct {
+	base     string
+	Verifier *ct.SignatureVerifier // nil for no verification (e.g. no public key available)
+	resolve  func(ctx context.Context, name string) ([]string, error)
+}
+
+// New constructs a new DNSClient instance.  The base parameter gives the
+// top-level domain name; opts can be used to provide a customer logger
+// interface and a public key for signature verification.
+func New(base string, opts jsonclient.Options) (*DNSClient, error) {
+	pubkey, err := opts.ParsePublicKey()
+	if err != nil {
+		return nil, fmt.Errorf("invalid public key: %v", err)
+	}
+
+	var verifier *ct.SignatureVerifier
+	if pubkey != nil {
+		var err error
+		verifier, err = ct.NewSignatureVerifier(pubkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(base) > 0 && base[len(base)-1] != '.' {
+		base += "."
+	}
+	return &DNSClient{
+		base:     base,
+		Verifier: verifier,
+		resolve:  func(ctx context.Context, name string) ([]string, error) { return net.LookupTXT(name) },
+	}, nil
+}
+
+const (
+	base64RE = "[A-Za-z0-9+/=]+"
+)
+
+var (
+	sthTXT = regexp.MustCompile(`^(\d+)\.(\d+)\.(` + base64RE + `)\.(` + base64RE + `)`)
+)
+
+// GetSTH retrieves the current STH from the log.
+func (c *DNSClient) GetSTH(ctx context.Context) (*ct.SignedTreeHead, error) {
+	name := "sth." + c.base
+	results, err := c.resolve(ctx, name)
+	if err != nil {
+		return nil, fmt.Errorf("lookup for %q failed: %v", name, err)
+	}
+	result := strings.Join(results, "")
+	matches := sthTXT.FindStringSubmatch(result)
+	if matches == nil || len(matches) < 5 {
+		return nil, fmt.Errorf("failed to parse result %q", result)
+	}
+	treeSize, err := strconv.ParseUint(matches[1], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse tree size %q", matches[1])
+	}
+	timestamp, err := strconv.ParseUint(matches[2], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse timestamp %q", matches[2])
+	}
+	rootHash, err := base64.StdEncoding.DecodeString(matches[3])
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse root hash %q", matches[3])
+	}
+	if len(rootHash) != sha256.Size {
+		return nil, fmt.Errorf("wrong size root hash %q", matches[3])
```

Comment:
Should include both actual and expected values.

Reply:
Done.